src/CircuitZoo/CircuitZoo.jl

```diff
@@ -743,4 +743,73 @@ function (circuit::PurifyExpedientNode)(purifiedL,sacrificedL)
     [a..., b..., c..., d...]
 end
 
+"""
+$TYPEDEF
+
+Fields:
+
+$FIELDS
+
+The circuit for Superdense Coding to encode the 2 (classical) bit message
+to its corresponding bell pair representation. `reg` is a single qubit register containing
+Alice's half of the entangled bell pair and message is the 2 bit message Alice intends to send to Bob.
+Based on the 2 bit message the state of the qubit in the register is mutated in place with `apply!`
+
+```jldoctest
+julia> reg = Register(1)
+
+julia> initialize!(reg[1], Z1)
+
+julia> message = [1, 1]
+
+julia> SDEncode()(reg, message)
+```
+"""
+
+struct SDEncode <: AbstractCircuit
+end
+
+function (circuit::SDEncode)(reg, message)
+    if message[1] == 1
+        apply!(reg[1], X)
+    end
+    if message[2] == 1
+        apply!(reg[1], Z)
+    end
+end
+
+"""
+$TYPEDEF
+
+Fields:
+
+$FIELDS
+
+The circuit for Superdense Coding to decode the 2 (classical) bit message
+using the entangled bell pair stored in the registers regA and regB after Alice's encoding of the first qubit.
+
+```jldoctest
+julia> regA = Register(1)
+
+julia> regB = Register(1)
+
+julia> initalize!(regA[1], Z1)
+
+julia> initalize!(regB[1], Z1)
+
+julia> SDDecode()(regA, regB)
+```
+"""
+
+struct SDDecode <: AbstractCircuit
+end
+
+function (circuit::SDDecode)(regA, regB)
+    apply!((regB[1], regA[1]), CNOT)
+    apply!(regB[1], H)
```

Comment:
Everything works fine upto this line, but when we apply H to Bob's qubit, the result is the same as what we would get if we applied H to Alice's qubit on paper, i.e., something like 
```julia
Ket(dim=4)
  basis: [Spin(1/2) ⊗ Spin(1/2)]
  0.49999999999999994 + 0.0im
 -0.49999999999999994 + 0.0im
 -0.49999999999999994 + 0.0im
  0.49999999999999994 + 0.0im
```
which gives us the random results.
But if we apply H to `regA[1]` , which is Alice's qubit here, it returns the correct results. I am not sure why it behaves this way, opposite to what we get when we solve it on paper, but it resolves the issue here

Reply:
Could you elaborate a bit? What is the state before the wrong step, what is the step that seems to break things, and what is the state after that step (and what should it be)?

Reply:
For the final part of the decoder circuit, we want to apply H to the second qubit (Bob's qubit), and H is a 1 qubit operator. But under the hood our register stores a 4 element ket vector which (if the classical message is `1 1`) would be:
```julia
Ket(dim=4)
  basis: [Spin(1/2) ⊗ Spin(1/2)]
                 0.0 + 0.0im
                 0.0 + 0.0im
  0.7071067811865475 + 0.0im
 -0.7071067811865475 + 0.0im
```
To apply the symbolic H to the second qubit, we call the `apply!` method at https://github.com/QuantumSavory/QuantumSavory.jl/blob/09fe3aa93d0daa4d0964feb606a795a1274ad566/src/baseops/apply.jl#L12
Here(https://github.com/QuantumSavory/QuantumSavory.jl/blob/09fe3aa93d0daa4d0964feb606a795a1274ad566/src/baseops/apply.jl#L21) the `state_indices` parameter of our ket determines whether the final operator would be `I⊗H`(correct) or `H⊗I`(wrong) given state_indices is 1 or 2 respectively. Since, Bob's qubit is the 2nd qubit of the subsystem `state_indices` is set to 2 and it finally hits the `apply!` methods in QuantumOpticsBase.jl. So, because of `state_indices` being 2, the final operator [here](https://github.com/qojulia/QuantumOpticsBase.jl/blob/fe00cb9fd15d867ccd37bdcf8ef9826da3149893/src/apply.jl#L2) (in QuantumOpticsBase.jl) becomes `H⊗I` which gives us the wrong answer in my previous comment.

So, [sorry for the huge description : ) ] to be short, the `embed` method [here](https://github.com/qojulia/QuantumOpticsBase.jl/blob/fe00cb9fd15d867ccd37bdcf8ef9826da3149893/src/apply.jl#L2) expects to be provided with the index where we want to put the identity operator(index 1) instead of the index of the qubit on which we're performing the single qubit operation(index 2), as described by its docstring [here](https://github.com/qojulia/QuantumOpticsBase.jl/blob/fe00cb9fd15d867ccd37bdcf8ef9826da3149893/src/operators.jl#L22)
Hence, I think we have a bug in this line
https://github.com/QuantumSavory/QuantumSavory.jl/blob/09fe3aa93d0daa4d0964feb606a795a1274ad566/src/baseops/apply.jl#L21
and `state_indices` should be a vector of all the indices of the state excluding the index of our qubit